torch/ao/sparsity/sparsifier.py

```diff
@@ -0,0 +1,182 @@
+            format_string += f'\tModule Group {i}\n'
+            format_string += f'\t    module: {module}\n'
+            for key in sorted(sparse_args.keys()):
+                if key == 'module':
+                    continue
+                format_string += f'\t    {key}: {sparse_args[key]}\n'
+        format_string += ')'
+        return format_string
+
+    def state_dict(self):
+        r"""Returns the state of the optimizer as a :class:`dict`.
+
+        It contains:
+        * module_groups - a list containing all sparsity configuration groups
+            with the key 'path' specifying the layer path within a model
+        """
+        module_groups = [
+            dict(filter(lambda key_value: key_value[0] != 'module', mg.items()))
+            for mg in self.module_groups
+        ]
+        return {
+            'module_groups': module_groups,
+        }
+
+    def load_state_dict(self, state_dict, strict=True):
+        module_groups = copy.deepcopy(state_dict['module_groups'])
+        for group in module_groups:
+            layer = _fqn_to_module(self.model, group['path'])
+            if strict and layer is None:
+                raise RuntimeError(f'Error loading group["path"] into the model')
+            group['module'] = layer
+        self.__setstate__({'module_groups': module_groups})
+
+    def prepare(self, model, config):
+        r"""Prepares a model, by adding the parametrizations.
+
+        Note::
+
+            The model is modified inplace. If you need to preserve the original
+            model, use copy.deepcopy.
+        """
+        self.config = config
+        self.model = model
+        # If no config -- try getting all the supported layers
+        if self.config is None:
+            # Add all models to the config
+            self.config = []
+            stack = [model]
+            while stack:
+                module = stack.pop()
+                for name, child in module.named_children():
+                    if type(child) in SUPPORTED_MODULES:
+                        self.config.append(child)
+                    else:
+                        stack.append(child)
+
+        for module_config in self.config:
+            if isinstance(module_config, nn.Module):
+                module_config = {'module': module_config}
+            local_args = copy.deepcopy(self.defaults)
+            local_args.update(module_config)
+            module = local_args['module']
+            module_path = _module_to_fqn(self.model, module)
+            if module_path and module_path[0] == '.':
+                module_path = module_path[1:]
+            local_args['path'] = module_path
+            self.module_groups.append(local_args)
+
+        self._prepare()
+
+    def _prepare(self, *args, **kwargs):
+        r"""Adds mask parametrization to the layer weight
+        """
+        for config in self.module_groups:
+            module = config['module']
+            if getattr(module, 'mask', None) is None:
```

Comment:
Can we have masks be only attr of FakeSparsity? If we change FakeSparsity's init to only take the weight shape, this is simpler. Currently there are two masks: one as an attr of the module and one as part of the FakeSparsityModule.
Suggest instead doing:
``` python
for config in self.module_groups:
            module = config['module']
            param = config.get('parametrization', FakeSparsity)
            parametrize.register_parametrization(module, 'weight',
                                                 param(getattr(module,'weight').shape))
```

Reply:
Oh the intention was that "if there is a mask in the module, it would use that, otherwise it would create one

Reply:
Lets make it explicit, the module should not have a mask, this way mask is always part of fakesparse module.